### ISSUE TITLE
IndexError: index 65655 is out of bounds for axis 0 with size 50176

### ISSUE THREAD

**Issue by iamani** [HuguesTHOMAS/KPConv-PyTorch#40]
Hi! I'm getting the following error from testing slam_segmentation/semanticKitti: 

e015-i0183 => L=0.626 acc= 79% / t(ms):   0.6  42.7  49.2)
```
Traceback (most recent call last):
  File "train_SemanticKitti.py", line 324, in <module>
    trainer.train(net, training_loader, test_loader, config)
  File "/KPConv-PyTorch/utils/trainer.py", line 167, in train
    for batch in training_loader:
  File "/anaconda3/envs/kpconv-pytorch/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 363, in __next__
    data = self._next_data()
  File "/anaconda3/envs/kpconv-pytorch/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 971, in _next_data
    return self._process_data(data)
  File "/anaconda3/envs/kpconv-pytorch/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 1014, in _process_data
    data.reraise()
  File "/anaconda3/envs/kpconv-pytorch/lib/python3.6/site-packages/torch/_utils.py", line 395, in reraise
    raise self.exc_type(msg)
IndexError: Caught IndexError in DataLoader worker process 6.
Original Traceback (most recent call last):
  File "/anaconda3/envs/kpconv-pytorch/lib/python3.6/site-packages/torch/utils/data/_utils/worker.py", line 185, in _worker_loop
    data = fetcher.fetch(index)
  File "/anaconda3/envs/kpconv-pytorch/lib/python3.6/site-packages/torch/utils/data/_utils/fetch.py", line 44, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/anaconda3/envs/kpconv-pytorch/lib/python3.6/site-packages/torch/utils/data/_utils/fetch.py", line 44, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/KPConv-PyTorch/datasets/SemanticKitti.py", line 311, in __getitem__
    p0 = new_points[wanted_ind, :3]
IndexError: index 65655 is out of bounds for axis 0 with size 50176
```

Can you please help? Is it error in the data? I am trying to reproduce the results of semantic kitti. Thanks!

**Comment by HuguesTHOMAS** [comment]
I dont recall seeing this error in my experiments. Did you use the code as is without any modification?

**Comment by iamani** [comment]
I only created a new datasets/config/dataloader of my custom datasets; I don't recall modifying any thing that may impact the available datasets; 

I pulled a new copy and had an issue similar to this and been fixed as suggested in the comment [issuecomment-662794917](https://github.com/HuguesTHOMAS/KPConv-PyTorch/issues/20#issuecomment-662794917)

But again the same error :-( 
````
e012-i0369 => L=0.653 acc= 78% / t(ms):   0.5  42.5  49.7)
Traceback (most recent call last):
  File "train_SemanticKitti.py", line 324, in <module>
    trainer.train(net, training_loader, test_loader, config)
  File "/new/KPConv-PyTorch/utils/trainer.py", line 167, in train
    for batch in training_loader:
  File "/anaconda3/envs/kpconv-pytorch/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 363, in __next__
    data = self._next_data()
  File "/anaconda3/envs/kpconv-pytorch/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 971, in _next_data
    return self._process_data(data)
  File "/anaconda3/envs/kpconv-pytorch/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 1014, in _process_data
    data.reraise()
  File "/anaconda3/envs/kpconv-pytorch/lib/python3.6/site-packages/torch/_utils.py", line 395, in reraise
    raise self.exc_type(msg)
IndexError: Caught IndexError in DataLoader worker process 6.
Original Traceback (most recent call last):
  File "/anaconda3/envs/kpconv-pytorch/lib/python3.6/site-packages/torch/utils/data/_utils/worker.py", line 185, in _worker_loop
    data = fetcher.fetch(index)
  File "/anaconda3/envs/kpconv-pytorch/lib/python3.6/site-packages/torch/utils/data/_utils/fetch.py", line 44, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/anaconda3/envs/kpconv-pytorch/lib/python3.6/site-packages/torch/utils/data/_utils/fetch.py", line 44, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/new/KPConv-PyTorch/datasets/SemanticKitti.py", line 311, in __getitem__
    p0 = new_points[wanted_ind, :3]
IndexError: index 81716 is out of bounds for axis 0 with size 50176
````

**Comment by HuguesTHOMAS** [comment]
Ok let me try to help you debug the problem.

Could you print the variable `self.balance_classes`, I think the value is True. Then can you print the shape of `sem_labels` which should be a np.array.

I assume the length of `sem_labels.shape` will be larger than the one of `new_points`, hence the error. If thats the case, you'll have to track the changes made to this two arrays (`sem_labels `and `new_points`). They should come from the same frame and thus have the same length. Try to print the name of the file they are loaded from. If they come from the same frame, but still have different number of points, then you might have a issue with the files you downloaded from SemanticKitti dataset.

**Comment by iamani** [comment]
Thanks @HuguesTHOMAS! It looks like in sequence 10 - scan 24 don't contain the same number of points as label 24 ... I will double check SematicKitti data files ...

**Comment by JohanBergius** [comment]
> Thanks @HuguesTHOMAS! It looks like in sequence 10 - scan 24 doesn't contain the same number of points as label 24 ... I will double check SematicKitti data files ...

Any updates?

**Comment by Yarroudh** [comment]
Hello,

I had the same issue when training with my own data. The dataset file I created was a copy of NPM3D with some changes made to .ply files prerpation function.

My dataset contains 17 labels, but I noticed that at I'm ignoring one class in this code:

```python
self.ignored_labels = np.array([0])
.
.
.
config.num_classes = self.num_classes - len(self.ignored_labels)
config.dataset_task = self.dataset_task
```

Even my dataset has classes numbers from 1 to 17, `len(self.ignored_labels)` is equal to 1 which results in `config.num_classes` equal to 16. So, just change `self.ignored_labels` = np.array([])` if you aren't ignoring any class, or write the correct number of the class.

I hope this was helpful.
@JohanBergius
